### PR TITLE
Closes #1802

### LINKED ISSUE #1802: Introduce separate class for warnings
## Example

```php
<?php
class Test extends PHPUnit_Framework_TestCase
{
}
```

## Actual

```
$ phpunit Test                                  
PHPUnit 5.0-g6caa139 by Sebastian Bergmann and contributors.

F

Time: 58 ms, Memory: 4.50Mb

There was 1 failure:

1) Warning
No tests found in class "Test".

FAILURES!
Tests: 1, Assertions: 0, Failures: 1.
```

## Expected

```
$ phpunit Test                                  
PHPUnit 5.0-g6caa139 by Sebastian Bergmann and contributors.

W

Time: 58 ms, Memory: 4.50Mb

There was 1 warning:

1) No tests found in class "Test".

WARNINGS!
Tests: 1, Assertions: 0, Warnings: 1.
```

### PR DESCRIPTION
**Pull request by Shaked**
Few comments: 
- I have added the `warning` as a separated thing than a `failure`. 
- The example in #1802 shows a very specific scenario for a warning, I have taken this a step farther as I saw that a warning can be shown from a different reasons. 
- As well, I kept the `1) Warning` title as I assumed that it was dropped by mistake. However, the fix for that is simple in case its required though I believe this will create a bit of a mess if thats true only for warnings while its not true for other scenarios. 

I hope this would do the job but feel free to comment out if anything comes in mind. 
Cheers
Shaked

### CODE REVIEW

File: src/Framework/TestListener.php [hunk 1]
```diff
@@ -25,6 +25,16 @@
     public function addError(PHPUnit_Framework_Test $test, Exception $e, $time);
 
     /**
+     * A warning occurred.
+     *
+     * @param PHPUnit_Framework_Test                 $test
+     * @param PHPUnit_Framework_AssertionWarning $e
+     * @param float                                  $time
+     * @since  Method available since Release 5.0.0
```
Comment (GrahamCampbell) [inline src/Framework/TestListener.php]: cs

File: src/Framework/TestResult.php [hunk 2]
```diff
@@ -626,6 +682,8 @@ function_exists('xdebug_start_function_monitor');
             } elseif ($e instanceof PHPUnit_Framework_SkippedTestError) {
                 $skipped = true;
             }
+        } catch(PHPUnit_Framework_AssertionWarning $e){
```
Comment (GrahamCampbell) [inline src/Framework/TestResult.php]: cs

**Comment by GrahamCampbell** [comment]
This fix looks nice.

File: src/Framework/TestResult.php [hunk 3]
```diff
@@ -244,6 +254,29 @@ public function addError(PHPUnit_Framework_Test $test, Exception $e, $time)
      * The passed in exception caused the failure.
      *
      * @param PHPUnit_Framework_Test                 $test
+     * @param PHPUnit_Framework_AssertionWarning $e
```
Comment (henriquemoody) [inline src/Framework/TestResult.php]: I think you forgot to update the copied dockblock, here, the current content is:

> Adds a failure to the list of failures.
> The passed in exception caused the failure.

I think you can just replace _failure_ by _warning_.

File: src/TextUI/ResultPrinter.php [hunk 4]
```diff
@@ -422,6 +452,20 @@ public function addFailure(PHPUnit_Framework_Test $test, PHPUnit_Framework_Asser
     }
 
     /**
+     * A failure occurred.
```
Comment (henriquemoody) [inline src/TextUI/ResultPrinter.php]: The same of my [last comment](https://github.com/sebastianbergmann/phpunit/pull/1838/files#r38142011).

**Comment by henriquemoody** [comment]
That's a great feature, thanks @Shaked! 

I think that would be awesome if you create 2 phpt tests for this feature:
- `stopOnWarning` configuration
- `--stop-on-warning` CLI argument

***
Ps.: There are a few Coding Standard that was not spotted by @GrahamCampbell, perhaps you should try to use PHPCS to check them.

**Comment by Shaked** [comment]
Thank you guys. I would definitely fix those issues now. 

I am originally using [php.tools](https://github.com/phpfmt/php.tools) and I canceled it as I wasn't sure about the code formatting. Maybe worth adding some information about the code formatting in the [CONTRIBUTING.md](https://github.com/sebastianbergmann/phpunit/blob/master/CONTRIBUTING.md) file, what do you think? (In case I haven't missed it). 

Shaked

**Comment by henriquemoody** [comment]
I agree that the [CONTRIBUTING.md](https://github.com/sebastianbergmann/phpunit/blob/master/CONTRIBUTING.md) file must be updated.
There are coding standards but seems that Sebastian Bergmann is very flexible with that, I haven't seem him requesting for CS fixed on pull requests, he fixes it himself, see a7df5a5, 3400886 and 9735163 -- I think it's a very good policy, BTW.

However, if you want you can try this: https://github.com/sebastianbergmann/phpunit/pull/1834#issuecomment-132896545

File: src/Framework/AssertionWarning.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Thrown when an assertion warns.
+ *
+ * @since Class available since Release 5.0.0
+ */
+class PHPUnit_Framework_AssertionWarning extends PHPUnit_Framework_Exception implements PHPUnit_Framework_SelfDescribing
```
Comment (sebastianbergmann) [inline src/Framework/AssertionWarning.php]: The name should be `PHPUnit_Framework_Warning`, a warning has nothing to do with an assertion.

**Comment by sebastianbergmann** [comment]
This needs to be retargetted for PHPUnit 5.1 (`master`).

**Comment by GrahamCampbell** [comment]
Shame we can't get this into 5.0. :(

File: src/Framework/AssertionWarning.php (hunk 5 above)
Comment (Shaked) [inline src/Framework/AssertionWarning.php]: Yes, I was thinking about it as a `warning` versus `error` and `failure` but I will change it. This means that the current `PHPUnit_Framework_Warning` will be changed to `PHPUnit_Framework_WarningTestCase`, does it make sense?

**Comment by Shaked** [comment]
I will move it to `master` (5.1). Thanks guys! :)

File: src/Framework/AssertionWarning.php (hunk 5 above)
Comment (sebastianbergmann) [inline src/Framework/AssertionWarning.php]: The point is/was to get rid of the existing `PHPUnit_Framework_Warning extends PHPUnit_Framework_TestCase` ugliness :-)

Comment (Shaked) [inline src/Framework/AssertionWarning.php]: Ok, I hope I understood it right, I will still use a WarningTestCase to know that a warning had happened. 

Lets continue in the upcoming pull request :) 
Thanks

**Comment by Shaked** [comment]
I have opened a new pull request https://github.com/sebastianbergmann/phpunit/pull/1840 to `master ` (5.1)

**Pull request closed without merging**